src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -0,0 +1,345 @@
+            // Finally this function needs to be called again to traverse all linked tables using the
+            // just found link.
+            find_all_related_tables(out, *table.get_link_target(col_key), tables_in_filters);
+        }
+    }
+    if (tables_in_filters.size() != 0) {
+        // Backlinks can only come into consideration when added via key paths.
+        // If there are not `tables_in_filter` we can skip this part.
+        table.for_each_backlink_column([&](ColKey column_key) {
+            auto target_table = table.get_link_target(column_key);
+            auto target_table_key = target_table->get_key();
+            if (any_of(tables_in_filters.begin(), tables_in_filters.end(), [target_table_key](auto table_key) {
+                    return target_table_key == table_key;
+                })) {
+                // If this backlink is included in any of the filters we follow the path further.
+                out[out_index].links.push_back({column_key.value, false});
+                find_all_related_tables(out, *table.get_link_target(column_key), tables_in_filters);
+            }
+            return false;
+        });
+    }
+}
+
+void DeepChangeChecker::find_filtered_related_tables(std::vector<RelatedTable>& out, Table const& table,
+                                                     std::vector<KeyPathArray> key_path_arrays)
+{
+    std::vector<TableKey> tables_in_filters = {};
+    for (auto key_path_array : key_path_arrays) {
+        for (auto key_path : key_path_array) {
+            for (auto key_path_element : key_path) {
+                tables_in_filters.push_back(key_path_element.first);
+            }
+        }
+    }
+    find_all_related_tables(out, table, tables_in_filters);
+}
+
+DeepChangeChecker::DeepChangeChecker(TransactionChangeInfo const& info, Table const& root_table,
+                                     std::vector<RelatedTable> const& related_tables,
+                                     std::vector<KeyPathArray> key_path_arrays)
+    : m_info(info)
+    , m_root_table(root_table)
+    , m_key_path_arrays(key_path_arrays)
+    , m_root_object_changes([&] {
+        auto it = info.tables.find(root_table.get_key().value);
+        return it != info.tables.end() ? &it->second : nullptr;
+    }())
+    , m_related_tables(related_tables)
+{
+    // If all callbacks do have a filter, every `KeyPathArray` will have entries.
+    // In this case we need to check the `ColKey`s and pass the filtered columns
+    // to the checker.
+    // If at least one `NotificationCallback` does not have a filter we notify on any change.
+    // This is signaled by leaving the `m_filtered_columns_in_root_table` and
+    // `m_filtered_columns` empty.
+    auto all_callbacks_filtered = all_of(begin(m_key_path_arrays), end(m_key_path_arrays), [](auto key_path_array) {
+        return key_path_array.size() > 0;
+    });
+    if (all_callbacks_filtered) {
+        for (auto key_path_array : m_key_path_arrays) {
```

Comment:
Same about `const auto&` when iterating over an array of strings or complex types.